src/ngamsPlugIns/ngamsPlugIns/alma/almaMirroringTargetNodeAssigner.py

```diff
@@ -58,7 +58,7 @@ def remove_hosts_without_available_threads(self, num_threads_per_host):
         """Could be that the hosts we are considering mirroring too already have theior hands full with mirroring. In
          theory each host has a limited number of threads which they should use for mirroring. We try not to exceed this
          Remove any hosts which do not have at least one thread available for mirroring"""
-        for next_host in self.hosts[:]:
+        for next_host in self.hosts:
```

Comment:
Are you sure you want to remove the list duplication here? 3 lines below you are removing `next_host` from `self.hosts`, so you probably want to iterate over a copy of `self.hosts`, hence the original `self.hosts[:]`

Reply:
Ah, I had not considered it was intentionally duplicating the list. Given that the following line `self.hosts.remove(next_host)` refers to the original `self.host` and not the duplicated list I would expect that it will modify the original list anyway. Perhaps it would be better to do something like...
```
available_host_list=list(self.hosts)
for next_host in available_host_list:
    ...
    available_host_list.remove(next-host)
...
return available_host_list
```
Do you think this is cleaner?

Reply:
Mmmm that's not doing the same as the original though. This last version differs from the original in that it doesn't change `self.hosts`, it changes the copied list. I haven't studied the context from which this code is invoked, so maybe the difference doesn't matter, but you're a better judge.

Reply:
Perhaps it copies the list but I think it still removes items from the original list. I tried the simple example...
```
#!/usr/bin/python2
tmp=[1,2,3,4,5,6,7,8,9]
for i in tmp[:]:                                                                                                                                          
    if i % 2:
        tmp.remove(i)
print tmp
```
I get the result `[2, 4, 6, 8]` which is what i would expect. The duplicated list makes no difference.
It could be that duplicating the list was incidental because the developer was not an experienced python programmer.

Reply:
That was just sheer luck actually :P

Try this:

```python
tmp = [1,2,3,4,5,6,7,8,9]
for i in tmp[:]:
    tmp.remove(i)
print(tmp)
```

Then remove the list duplication and see `tmp` not have what you'd expect.

If the intent of the original code is to remove elements from `self.hosts` then you'll need the list duplication for iteration purposes.

Reply:
I see your point. Jeez, my python skills are pretty bad. I will restore the `[:]` for now. Although I feel the object should maintain the default list and then the method can be reused. I think a new object is created every time and this is how a fresh list of hosts is obtained.
Although the comments in the code are strange.
```
        for next_host in self.hosts[:]:
            if self.num_files_to_be_mirrored[next_host] >= num_threads_per_host:
                logger.info('host %s does not have any threads available for mirroring - removing as a target', next_host)
                self.hosts.remove(next_host)
        logger.info('remaining target nodes %r', self.hosts)
        # now we need to update the start host index

        # return is for unit testing
        return self.hosts
```
The return value is used by the mirroring plugin therefore it is not only for unit tests. The previous comment appears to be redundant. The mirroring code is confusing to say the least.

Reply:
> I see your point. Jeez, my python skills are pretty bad. I will restore the [:] for now.

Yeap, I saw the change now :+1:. I remember that Java throws as `ConcurrentModificationException` when you modify while iterating, making it easy to spot these occurrences. Python OTOH doesn't, so it's way more difficult to see that this is an illegal operation, and unless you pay close attention you might end up with results you think are correct but aren't.